libraries/AP_HAL_ChibiOS/UARTDriver.cpp

```
@@ -83,51 +97,113 @@ _baudrate(57600)
+            need_tick = true;
             last_thread_run_us = now;
         }
+
+        // change the thread priority if requested - if unbuffered it should only have higher priority than the owner so that
+        // handoff occurs immediately
+        if (mask & EVT_TRANSMIT_UNBUFFERED) {
+            chThdSetPriority(unbuffered_writes ? MIN(_uart_owner_thd->prio + 1, APM_UART_UNBUFFERED_PRIORITY) : APM_UART_PRIORITY);
+        }
+#ifndef HAL_UART_NODMA
+        osalDbgAssert(!dma_handle || !dma_handle->is_locked(), "DMA handle is already locked");
+#endif
+        // send more data
+        if (_tx_initialised && ((mask & EVT_TRANSMIT_DATA_READY) || need_tick || (hd_tx_active && (mask & EVT_TRANSMIT_END)))) {
+            _tx_timer_tick();
+        }
+    }
+}
+#pragma GCC diagnostic pop
+
+/*
+  thread for handling UART receive
+
+  We use events indexed by serial_num to trigger a more rapid send for
+  unbuffered_write uarts, and run at 1kHz for general UART handling
+ */
+void UARTDriver::uart_rx_thread(void* arg)
+{
+    while (uart_rx_thread_ctx == nullptr) {
+        hal.scheduler->delay_microseconds(1000);
+    }
+
+    while (true) {
+        hal.scheduler->delay_microseconds(1000);
+
         for (uint8_t i=0; i<UART_MAX_DRIVERS; i++) {
             if (uart_drivers[i] == nullptr) {
                 continue;
             }
-            if (uart_drivers[i]->_initialised &&
-                (mask & EVENT_MASK(i) ||
-                 (uart_drivers[i]->hd_tx_active && (mask & EVT_TRANSMIT_END)))) {
-                uart_drivers[i]->_timer_tick();
+            if (uart_drivers[i]->_rx_initialised) {
+                uart_drivers[i]->_rx_timer_tick();
             }
         }
     }
 }
 
 /*
-  initialise UART thread
+  initialise UART RX thread
  */
-void UARTDriver::thread_init(void)
+void UARTDriver::thread_rx_init(void)
 {
-    if (uart_thread_ctx) {
-        // already initialised
-        return;
+#if CH_CFG_USE_HEAP == TRUE
```

better to not have this #if - we'd rather not compile than silently fail to work. We need heap for everything other than bootloader, and the bootloader sets HAL_NO_UARTDRIVER